Base/QTCore/qSlicerExtensionsManagerModel.cxx

```diff
@@ -2038,35 +1999,47 @@ bool qSlicerExtensionsManagerModel::updateExtensionsMetadataFromServer(bool forc
         return false;
         }
       d->ExtensionsMetadataFromServerAPI.setServerUrl(this->serverUrl().toString() + QString("/api/v1/app/%1/extension").arg(appID));
+      parameters["app_revision"] = this->slicerRevision();
+      parameters["os"] = this->slicerOs();
+      parameters["arch"] = this->slicerArch();
+      // request all metadata in a single response (it makes synchronous query simpler)
+      parameters["limit"] = QString::number(0);
       }
     else
       {
       qWarning() << "Update extension information from server failed: missing implementation for serverAPI" << this->serverAPI();
       return false;
       }
-    d->queryExtensionsMetadataFromServer();
+
+    // Issue the query
+    d->ExtensionsMetadataFromServerQueryUID = d->ExtensionsMetadataFromServerAPI.get("", parameters);
     }
 
   if (!waitForCompletion)
     {
     return true;
     }
 
-  // Wait up to 15 seconds for the update check to complete
-  // (typically takes only a few seconds)
-  int timeoutMsec = 15000;
-  bool completedSignalReceived = spy.wait(timeoutMsec);
-  if (!completedSignalReceived)
+  // Temporarily disable onExtensionsMetadataFromServerQueryFinished call via signal/slot
+  // because we'll call it directly to get returned result.
+  QObject::disconnect(&d->ExtensionsMetadataFromServerAPI,
+    SIGNAL(finished(QUuid)),
+    this, SLOT(onExtensionsMetadataFromServerQueryFinished(QUuid)));
```

Comment:
Instead of disconnect/connect, instantiating and using `qGirderAPI` and conditionally calling `.sync()` would also work.

Since until recently, we were supporting both midas and girder, the querying was done using the more generic `qRestAPI` along with explicit handling of the parsing using the static helper function like `qGirderAPI::parseGirderAPIv1Response`.

Now that we fully transitioned, using an instance of `qGirderAPI` would ensure the virtual function `qGirderAPI::parseResponse` is used ....

```diff
- qRestAPI ExtensionsMetadataFromServerAPI;
+ qGirderAPI ExtensionsMetadataFromServerAPI;
```

Reply:
This could also be done at a later time as this PR addresses the problem.

Reply:
I've tried using sync but there were two errors:
1. If the `finished()` signal is not disconnected then `onExtensionsMetadataFromServerQueryFinished` is called (which takes the result) and that makes `sync()` crash.
2. Sync cannot return a `qRestResult`, only a `QList<QVariantMap>`, so I would have needed to implement two different result parsing mechanisms (one for sync and another one for async).

`takeResult` waits for the result anyway (by calling `waitForDone()`) therefore using `sync()` would have just complicated things.